File: archiso/airootfs/usr/share/churros/preferences/widgets/group.py
import gi

gi.require_version("Gtk", "4.0")

from gi.repository import Gtk


class Group(Gtk.Box):

    def __init__(
        self,
        title
    ):

        super().__init__(
            orientation=Gtk.Orientation.VERTICAL,
            spacing=10
        )

        self.set_margin_bottom(24)

        #
        # Título
        #

        label = Gtk.Label(
            label=title
        )

        label.set_xalign(0)

        label.add_css_class(
            "group-title"
        )

        self.append(
            label
        )

        #
        # Tarjeta
        #

        self.card = Gtk.Box(
            orientation=Gtk.Orientation.VERTICAL
        )

        self.card.add_css_class(
            "group-card"
        )

        self.append(
            self.card
        )

        self.first = True

    def clear(
        self
    ):

        child = self.card.get_first_child()

        while child is not None:

            next_child = child.get_next_sibling()

            self.card.remove(
                child
            )

            child = next_child

        self.first = True

    def add(
        self,
        widget
    ):

        if not self.first:

            separator = Gtk.Separator(
                orientation=Gtk.Orientation.HORIZONTAL
            )

            separator.add_css_class(
                "group-separator"
            )

            self.card.append(
                separator
            )

        self.card.append(
            widget
        )

        self.first = False

File: rust/preferences/src/widgets/group.rs
// ==========================================
// Group — tarjeta con título y separadores entre filas
// (equivalente a widgets/group.py)
// ==========================================

use gtk::prelude::*;

use crate::widgets::AsWidget;

pub struct Group {
    pub root: gtk::Box,
    pub card: gtk::Box,
    first: bool,
}

impl Group {
    pub fn new(title: &str) -> Self {
        let root = gtk::Box::new(gtk::Orientation::Vertical, 10);
        root.set_margin_bottom(24);

        let label = gtk::Label::builder()
            .label(title)
            .xalign(0.0)
            .build();
        label.add_css_class("group-title");
        root.append(&label);

        let card = gtk::Box::new(gtk::Orientation::Vertical, 0);
        card.add_css_class("group-card");
        root.append(&card);

        Self {
            root,
            card,
            first: true,
        }
    }

    pub fn widget(&self) -> &gtk::Box {
        &self.root
    }

    pub fn clear(&mut self) {
        while let Some(child) = self.card.first_child() {
            self.card.remove(&child);
        }
        self.first = true;
    }

    pub fn add(&mut self, widget: &impl AsWidget) {
        if !self.first {
            let separator = gtk::Separator::new(gtk::Orientation::Horizontal);
            separator.add_css_class("group-separator");
            self.card.append(&separator);
        }
        self.card.append(widget.widget());
        self.first = false;
    }
}
